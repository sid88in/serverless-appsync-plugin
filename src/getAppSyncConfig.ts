import {
  AppSyncConfig,
  isSharedApiConfig,
  PipelineResolverConfig,
  UnitResolverConfig,
} from './types/index.js';

import type {
  ApiKeyConfig,
  AppSyncConfig as PluginAppSyncConfig,
  DataSourceConfig,
  PipelineFunctionConfig,
  ResolverConfig,
  BaseAppSyncConfig,
  SharedAppSyncConfig,
  FullAppSyncConfig,
  Substitutions,
} from './types/plugin.js';
import { forEach, merge } from 'lodash-es';

const flattenMaps = <T>(
  input?: Record<string, T> | Record<string, T>[],
): Record<string, T> => {
  if (Array.isArray(input)) {
    return merge({}, ...input);
  } else {
    return merge({}, input);
  }
};

export const isUnitResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is UnitResolverConfig => {
  return resolver.kind === 'UNIT';
};

export const isPipelineResolver = (resolver: {
  kind?: 'UNIT' | 'PIPELINE';
}): resolver is PipelineResolverConfig => {
  return !resolver.kind || resolver.kind === 'PIPELINE';
};

const toResourceName = (name: string) => {
  return name.replace(/[^a-z_]/i, '_');
};

export const getAppSyncConfig = (
  config: AppSyncConfig,
): PluginAppSyncConfig => {
  const baseConfig = getBaseAppsyncConfig(config);

  // handle shared appsync config
  if (isSharedApiConfig(config)) {
    const apiId: string = config.apiId;
    return {
      ...baseConfig,
      apiId,
    } satisfies SharedAppSyncConfig;
  }

  // Handle full appsync config
  const schema = Array.isArray(config.schema)
    ? config.schema
    : [config.schema || 'schema.graphql'];

  const additionalAuthentications = config.additionalAuthentications || [];

  let apiKeys: Record<string, ApiKeyConfig> | undefined;
  if (
    config.authentication?.type === 'API_KEY' ||
    additionalAuthentications.some((auth) => auth.type === 'API_KEY')
  ) {
    const inputKeys = config.apiKeys || [];

    apiKeys = inputKeys.reduce((acc, key) => {
      if (typeof key === 'string') {
        acc[key] = { name: key };
      } else {
        acc[key.name] = key;
      }

      return acc;
    }, {});
  }

  return {
    ...config,
    ...baseConfig,
    additionalAuthentications,
    apiKeys,
    schema,
  } satisfies FullAppSyncConfig;
};

function getBaseAppsyncConfig(config: AppSyncConfig): BaseAppSyncConfig {
  const dataSources: Record<string, DataSourceConfig> = {};
  const resolvers: Record<string, ResolverConfig> = {};
  const pipelineFunctions: Record<string, PipelineFunctionConfig> = {};
  const substitutions: Substitutions = {};

  forEach(flattenMaps(config.dataSources), (ds, name) => {
    dataSources[name] = {
      ...ds,
      name,
    };
  });

  forEach(flattenMaps(config.resolvers), (resolver, typeAndField) => {
    const [type, field] = typeAndField.split('.');

    if (typeof resolver === 'string') {
      resolvers[typeAndField] = {
        dataSource: resolver,
        kind: 'UNIT',
        type,
        field,
      };
      return;
    }

    if (isUnitResolver(resolver) && typeof resolver.dataSource === 'object') {
      const name = typeAndField.replace(/[^a-z_]/i, '_');
      dataSources[name] = {
        ...resolver.dataSource,
        name,
      };
    }

    resolvers[typeAndField] = {
      ...resolver,
      type: resolver.type || type,
      field: resolver.field || field,
      ...(isUnitResolver(resolver)
        ? {
            kind: 'UNIT',
            dataSource:
              typeof resolver.dataSource === 'object'
                ? typeAndField.replace(/[^a-z_]/i, '_')
                : resolver.dataSource,
          }
        : {
            kind: 'PIPELINE',
            functions: resolver.functions.map((f, index) => {
              if (typeof f === 'string') {
                return f;
              }

              const name = `${toResourceName(typeAndField)}_${index}`;
              pipelineFunctions[name] = {
                ...f,
                name,
                dataSource:
                  typeof f.dataSource === 'string' ? f.dataSource : name,
              };
              if (typeof f.dataSource === 'object') {
                dataSources[name] = {
                  ...f.dataSource,
                  name,
                };
              }
              return name;
            }),
          }),
    };
  });

  forEach(flattenMaps(config.pipelineFunctions), (func, name) => {
    if (typeof func.dataSource === 'object') {
      dataSources[name] = {
        ...func.dataSource,
        name,
      };
    }

    pipelineFunctions[name] = {
      ...func,
      dataSource: typeof func.dataSource === 'string' ? func.dataSource : name,
      name,
    };
  });

  if (config.substitutions) {
    Object.assign(substitutions, config.substitutions);
  }

  return {
    dataSources,
    resolvers,
    pipelineFunctions,
    substitutions,
  };
}
