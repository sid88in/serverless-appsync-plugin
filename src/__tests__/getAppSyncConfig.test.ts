import { pick } from 'lodash-es';
import { getAppSyncConfig } from '../getAppSyncConfig.js';
import { basicConfig } from './basicConfig.js';
import { ResolverConfig } from '../types/index.js';

test('returns basic config', async () => {
  expect(getAppSyncConfig(basicConfig)).toMatchSnapshot();
});

describe('Schema', () => {
  it('should return the default schema', () => {
    const config = getAppSyncConfig({ ...basicConfig, schema: undefined });
    const schema = 'schema' in config ? config.schema : undefined;
    expect(schema).toMatchSnapshot();
  });

  it('should return a single schema as an array', () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      schema: 'mySchema.graphql',
    });
    const schema = 'schema' in config ? config.schema : undefined;
    expect(schema).toMatchSnapshot();
  });

  it('should return a schema array unchanged', () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      schema: ['users.graphql', 'posts.graphql'],
    });
    const schema = 'schema' in config ? config.schema : undefined;
    expect(schema).toMatchSnapshot();
  });
});

describe('Api Keys', () => {
  it('should not generate a default Api Key when auth is not API_KEY', () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      authentication: { type: 'AWS_IAM' },
    });
    const apiKeys = 'apiKeys' in config ? config.apiKeys : undefined;
    expect(apiKeys).toBeUndefined();
  });

  it('should generate api keys', () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      apiKeys: [
        {
          name: 'John',
          description: "John's key",
          expiresAt: '2021-03-09T16:00:00+00:00',
        },
        {
          name: 'Jane',
          expiresAfter: '1y',
        },
        'InlineKey',
      ],
    });
    const apiKeys = 'apiKeys' in config ? config.apiKeys : undefined;

    expect(apiKeys).toMatchInlineSnapshot(`
      Object {
        "InlineKey": Object {
          "name": "InlineKey",
        },
        "Jane": Object {
          "expiresAfter": "1y",
          "name": "Jane",
        },
        "John": Object {
          "description": "John's key",
          "expiresAt": "2021-03-09T16:00:00+00:00",
          "name": "John",
        },
      }
    `);
  });
});

describe('DataSources', () => {
  it('should merge dataSource arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      dataSources: [
        {
          myDataSource: {
            type: 'NONE',
          },
          myOtherDataSource: {
            type: 'NONE',
          },
        },
        {
          otherSource: {
            type: 'NONE',
          },
          anotherNamedSource: {
            type: 'NONE',
          },
        },
      ],
    });
    expect(config.dataSources).toMatchSnapshot();
  });

  it('should merge dataSources embedded into resolvers and pipelineFunctions', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      dataSources: {
        myDataSource: {
          type: 'NONE',
        },
        myOtherDataSource: {
          type: 'NONE',
        },
      },
      resolvers: {
        'Query.getUser': {
          kind: 'UNIT',
          dataSource: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'getUser',
            },
          },
        },
        getUsers: {
          kind: 'UNIT',
          type: 'Query',
          field: 'getUsers',
          dataSource: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'getUsers',
            },
          },
        },
        'Mutation.createUser': {
          kind: 'PIPELINE',
          functions: [
            {
              dataSource: {
                type: 'AWS_LAMBDA',
                config: {
                  functionName: 'createUser',
                },
              },
            },
          ],
        },
      },
      pipelineFunctions: {
        function1: {
          dataSource: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'function1',
            },
          },
        },
        function2: {
          dataSource: {
            type: 'AWS_LAMBDA',
            config: {
              functionName: 'function2',
            },
          },
        },
      },
    });
    expect(
      pick(config, ['dataSources', 'resolvers', 'pipelineFunctions']),
    ).toMatchSnapshot();
  });
});

describe('Resolvers', () => {
  it('should resolve resolver type and fields', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      resolvers: {
        'Query.getUser': {
          kind: 'UNIT',
          dataSource: 'users',
        },
        getUsersResolver: {
          type: 'Query',
          field: 'getUsers',
          kind: 'UNIT',
          dataSource: 'users',
        },
      },
    });
    expect(config.resolvers).toMatchSnapshot();
  });

  it('should merge resolvers arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      resolvers: [
        {
          'Query.getUser': {
            kind: 'UNIT',
            dataSource: 'users',
          },
          getUsersResolver: {
            kind: 'UNIT',
            type: 'Query',
            field: 'getUsers',
            dataSource: 'users',
          },
          'Query.pipeline': {
            kind: 'PIPELINE',
            functions: ['function1', 'function2'],
          },
        },
        {
          'Query.getPost': {
            kind: 'UNIT',
            dataSource: 'posts',
          },
          getPostsResolver: {
            type: 'Query',
            kind: 'UNIT',
            field: 'getPosts',
            dataSource: 'posts',
          },
          pipelineResolver2: {
            kind: 'PIPELINE',
            functions: ['function1', 'function2'],
            type: 'Query',
            field: 'getUsers',
          },
        },
      ] satisfies Record<string, ResolverConfig>[],
    });
    expect(config.resolvers).toMatchSnapshot();
  });
});

describe('Pipeline Functions', () => {
  it('should merge function arrays', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,

      pipelineFunctions: [
        {
          function1: {
            dataSource: 'users',
          },
          function2: {
            dataSource: 'users',
          },
        },
        {
          function3: {
            dataSource: 'users',
          },
          function4: {
            dataSource: 'users',
          },
        },
      ],
    });
    expect(config.pipelineFunctions).toMatchSnapshot();
  });

  it('should merge inline function definitions', async () => {
    const config = getAppSyncConfig({
      ...basicConfig,
      resolvers: {
        'Mutation.createUser': {
          kind: 'PIPELINE',
          functions: [
            {
              dataSource: 'users',
              code: 'function1.js',
            },
            {
              dataSource: 'users',
              code: 'function2.js',
            },
          ],
        },
        'Mutation.updateUser': {
          kind: 'PIPELINE',
          functions: [
            {
              code: 'updateUser.js',
              dataSource: {
                type: 'AWS_LAMBDA',
                config: {
                  functionName: 'updateUser',
                },
              },
            },
          ],
        },
      },
      pipelineFunctions: {
        function1: {
          dataSource: 'users',
        },
        function2: {
          dataSource: 'users',
        },
      },
    });
    expect(config.pipelineFunctions).toMatchSnapshot();
  });
});
