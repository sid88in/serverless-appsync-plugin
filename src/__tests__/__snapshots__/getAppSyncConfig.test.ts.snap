// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSources should merge dataSource arrays 1`] = `
Object {
  "anotherNamedSource": Object {
    "name": "anotherNamedSource",
    "type": "NONE",
  },
  "myDataSource": Object {
    "name": "myDataSource",
    "type": "NONE",
  },
  "myOtherDataSource": Object {
    "name": "myOtherDataSource",
    "type": "NONE",
  },
  "otherSource": Object {
    "name": "otherSource",
    "type": "NONE",
  },
}
`;

exports[`DataSources should merge dataSources embedded into resolvers and pipelineFunctions 1`] = `
Object {
  "dataSources": Object {
    "Query_getUser": Object {
      "config": Object {
        "functionName": "getUser",
      },
      "name": "Query_getUser",
      "type": "AWS_LAMBDA",
    },
    "function1": Object {
      "config": Object {
        "functionName": "function1",
      },
      "name": "function1",
      "type": "AWS_LAMBDA",
    },
    "function2": Object {
      "config": Object {
        "functionName": "function2",
      },
      "name": "function2",
      "type": "AWS_LAMBDA",
    },
    "getUsers": Object {
      "config": Object {
        "functionName": "getUsers",
      },
      "name": "getUsers",
      "type": "AWS_LAMBDA",
    },
    "myDataSource": Object {
      "name": "myDataSource",
      "type": "NONE",
    },
    "myOtherDataSource": Object {
      "name": "myOtherDataSource",
      "type": "NONE",
    },
  },
  "resolvers": Object {
    "Query.getUser": Object {
      "dataSource": "Query_getUser",
      "field": "getUser",
      "kind": "UNIT",
      "type": "Query",
    },
    "getUsers": Object {
      "dataSource": "getUsers",
      "field": "getUsers",
      "kind": "UNIT",
      "type": "Query",
    },
  },
}
`;

exports[`Pipeline Functions should merge function arrays 1`] = `
Object {
  "function1": Object {
    "dataSource": "users",
    "name": "function1",
  },
  "function2": Object {
    "dataSource": "users",
    "name": "function2",
  },
  "function3": Object {
    "dataSource": "users",
    "name": "function3",
  },
  "function4": Object {
    "dataSource": "users",
    "name": "function4",
  },
  "function5": Object {
    "dataSource": "users",
    "name": "function5",
  },
}
`;

exports[`Resolvers should merge resolvers arrays 1`] = `
Object {
  "Query.getComments": Object {
    "dataSource": "comments",
    "field": "getComments",
    "kind": "UNIT",
    "type": "Query",
  },
  "Query.getPost": Object {
    "dataSource": "posts",
    "field": "getPost",
    "kind": "UNIT",
    "type": "Query",
  },
  "Query.getUser": Object {
    "dataSource": "users",
    "field": "getUser",
    "kind": "UNIT",
    "type": "Query",
  },
  "Query.pipeline": Object {
    "field": "pipeline",
    "functions": Array [
      "function1",
      "function2",
    ],
    "kind": "PIPELINE",
    "type": "Query",
  },
  "getPostsResolver": Object {
    "dataSource": "posts",
    "field": "getPosts",
    "kind": "UNIT",
    "type": "Query",
  },
  "getUsersResolver": Object {
    "dataSource": "users",
    "field": "getUsers",
    "kind": "UNIT",
    "type": "Query",
  },
  "pipelineResolver2": Object {
    "field": "getUsers",
    "functions": Array [
      "function1",
      "function2",
    ],
    "kind": "PIPELINE",
    "type": "Query",
  },
}
`;

exports[`Resolvers should resolve resolver type and fields 1`] = `
Object {
  "Query.getPosts": Object {
    "dataSource": "posts",
    "field": "getPosts",
    "kind": "UNIT",
    "type": "Query",
  },
  "Query.getUser": Object {
    "dataSource": "users",
    "field": "getUser",
    "kind": "UNIT",
    "type": "Query",
  },
  "getUsersResolver": Object {
    "dataSource": "users",
    "field": "getUsers",
    "kind": "UNIT",
    "type": "Query",
  },
}
`;

exports[`Schema should return a schema array unchanged 1`] = `
Array [
  "users.graphql",
  "posts.graphql",
]
`;

exports[`Schema should return a single schema as an array 1`] = `
Array [
  "mySchema.graphql",
]
`;

exports[`Schema should return the default schema 1`] = `
Array [
  "schema.graphql",
]
`;

exports[`returns basic config 1`] = `
Object {
  "additionalAuthentications": Array [],
  "apiKeys": Array [],
  "authentication": Object {
    "type": "API_KEY",
  },
  "dataSources": Object {},
  "mappingTemplatesLocation": Object {
    "pipelineFunctions": "mapping-templates",
    "resolvers": "mapping-templates",
  },
  "name": "My Api",
  "pipelineFunctions": Object {},
  "resolvers": Object {},
  "schema": Array [
    "schema.graphql",
  ],
}
`;
